Ext4Pkg/Ext4Dxe/File.c

```diff
@@ -134,13 +134,230 @@ Ext4DirCanLookup (
+  if (!EFI_ERROR (Status)) {
+    File->Position += SymlinkSizeTmp;
+  } else {
+    DEBUG ((DEBUG_FS, "[ext4] Failed to read symlink from blocks with Status %r\n", Status));
+    FreePool (SymlinkTmp);
+    return Status;
+  }
+
+  //
+  // Add null-terminator
+  //
+  SymlinkTmp[SymlinkSizeTmp] = '\0';
+
+  if (SymlinkSizeTmp != EXT4_INODE_SIZE (File->Inode)
+      || SymlinkAllocateSize != AsciiStrSize (SymlinkTmp)) {
+    Status = EFI_VOLUME_CORRUPTED;
+    DEBUG ((
+      DEBUG_FS,
+      "[ext4] Error! The sz of the read block doesn't match the value from the inode!\n"
+      ));
+    return Status;
+  }
+
+  *AsciiSymlinkSize = SymlinkSizeTmp;
+  *AsciiSymlink = SymlinkTmp;
+
+  Status = EFI_SUCCESS;
+  return Status;
+}
+
+/**
+  Reads a symlink file.
+
+  @param[in]      Partition   Pointer to the ext4 partition.
+  @param[in]      File        Pointer to the open symlink file.
+  @param[out]     Symlink     Pointer to the output unicode symlink string.
+
+  @retval EFI_SUCCESS           Symlink was read.
+  @retval EFI_ACCESS_DENIED     Symlink is encrypted.
+  @retval EFI_OUT_OF_RESOURCES  Memory allocation error.
+  @retval EFI_INVALID_PARAMETER Symlink path has incorrect length
+  @retval EFI_VOLUME_CORRUPTED  Symlink read block size differ from inode value
+**/
+EFI_STATUS
+Ext4ReadSymlink (
+  IN     EXT4_PARTITION  *Partition,
+  IN     EXT4_FILE       *File,
+  OUT    CHAR16          **Symlink
+  )
+{
+  EFI_STATUS  Status;
+  CHAR8       *SymlinkTmp;
+  UINTN       SymlinkSize;
+  UINTN       SymlinkAllocateSize;
+  CHAR16      *Symlink16Tmp;
+  CHAR16      *Needle;
+
+  //
+  // Assume that we alread read Inode via Ext4ReadInode
+  // Skip reading, just check encryption flag
+  //
+  if ((File->Inode->i_flags & EXT4_ENCRYPT_FL) != 0) {
+    Status = EFI_ACCESS_DENIED;
+    DEBUG ((DEBUG_FS, "[ext4] Error, symlink is encrypted\n"));
+    return Status;
+  }
+
+  if (Ext4SymlinkIsFastSymlink(File)) {
+    Status = Ext4ReadFastSymlink (Partition, File, &SymlinkTmp, &SymlinkSize);
+  } else {
+    Status = Ext4ReadSlowSymlink (Partition, File, &SymlinkTmp, &SymlinkSize);
+  }
+
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_FS, "[ext4] Symlink read error with Status %r\n", Status));
+    return Status;
+  }
+
+  SymlinkAllocateSize = SymlinkSize + 1;
+
+  Symlink16Tmp = AllocateZeroPool (SymlinkAllocateSize * sizeof (CHAR16));
+  if (Symlink16Tmp == NULL) {
+    Status = EFI_OUT_OF_RESOURCES;
+    DEBUG ((DEBUG_FS, "[ext4] Failed to allocate symlink unicode string buffer\n"));
+    FreePool (SymlinkTmp);
+    return Status;
+  }
+
+  Status = AsciiStrToUnicodeStrS (
+    SymlinkTmp,
+    Symlink16Tmp,
+    SymlinkSize * sizeof (CHAR16)
+    );
+
+  if (EFI_ERROR (Status)) {
+    DEBUG ((
+      DEBUG_FS,
+      "[ext4] Failed to convert ascii symlink to unicode with Status %r\n",
+      Status
+      ));
+    FreePool (SymlinkTmp);
+    FreePool (Symlink16Tmp);

```

Comment:
This is nitpicking of course, but try to free in the reverse order of allocation. This both makes logical sense and it allows the compiler to re-use exits (e.g. it may free SymlinkTmp in-place and then jump to a common exit with the branch above to free SymlinkTmp.

Reply:
Ok, https://github.com/acidanthera/audk/pull/11/commits/7d70fb2e8aef5187312cc05fd8dda4ebc0cba1db